stream-core/src/main/java/org/apache/kylin/stream/core/query/MultiThreadsResultCollector.java

```diff
@@ -69,38 +81,41 @@ public MultiThreadsResultCollector(int numOfWorkers, int timeout) {
 
             @Override
             public boolean hasNext() {
-                boolean exits = (internalIT.hasNext() || queue.size() > 0);
+                boolean exits = (internalIT.hasNext() || !recordCachePool.isEmpty());
                 if (!exits) {
                     while (notCompletedWorkers.get() > 0) {
                         Thread.yield();
-                        long takeTime = System.currentTimeMillis() - startTime;
-                        if (takeTime > timeout) {
+                        if (System.currentTimeMillis() > deadline) {
+                            masterThread.interrupt(); // notify main thread
```

Comment:
Will the workersSemaphore be released if masterThread is interrupted?

Reply:
1. If `WorkSubmitter` is interrupted, `WorkSubmitter` will leave it `run` method. As you can see in finally block of `run` method, it cancel every sub-thread(`ResultIterateWorker` will thrown InterruptedException).

2. In finally block of `ResultIterateWorker` 's `run`, it will release workersSemaphore.

Reply:
OK, I see it